### PR TITLE
Do not revalidate encrypted attribute with current encrypted_type

### LINKED ISSUE #48248: Rails 7.0: Double validation error for `validates_uniqueness_of` when attribute is encrypted
### Steps to reproduce
In the `User` model, I have an attribute (`User.thing`) that is encrypted and has a uniqueness validation:

``` ruby
class User < ApplicationRecord

  encrypts :thing, deterministic: true, downcase: true
  
  validates :thing, uniqueness: true

end
```

When I attempt to trigger this validation error (by submitting a `User` with a value for this attribute that already exists in a different `User`), the uniqueness error message is displayed 2 times, instead of only once.

When I output and inspect this `User`'s errors in the log file, I see that 1 error message contains the **unencrypted** value, and 1 additional error messages are added for the **encrypted** value.

It looks like this:

``` ruby
#<ActiveModel::Errors [

#<ActiveModel::Error attribute=thing, type=taken, options={:value=>"my_test_string"}>, 
#<ActiveModel::Error attribute=thing, type=taken, options={:value=>"{\"p\":\"35Q9AxGPf1w9uWS3lAkbXZULHiU=\",\"h\":{\"iv\":\"XYR33B22apwOsVwP\",\"at\":\"on8mfHa5ePMUCgVjjn81oQ==\"}}"}>

]>
```
### Actual behavior
2 "duplicate" error messages are generated/displayed.

### Expected behavior
Only 1 error message should be generated/displayed.

### System configuration
- Ruby **3.0.1p64**
- Rails **7.0.4.3**

### LINKED ISSUE #48445: Additional Query incase of Uniqueness validation for Encrypted Attribute
### Steps to reproduce

- Set Deterministic option to true for an Encrypted attribute
- Add uniqueness validation for the Same Encrypted attribute

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # gem "rails", github: "rails/rails", branch: "main"
  gem 'rails', '~> 7.0.4.3'
  gem "sqlite3"
  gem "pry"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Encryption.configure(
  primary_key: SecureRandom.alphanumeric(32),
  deterministic_key: SecureRandom.alphanumeric(32),
  key_derivation_salt: SecureRandom.alphanumeric(32),
  extend_queries: true,
)

ActiveRecord::Encryption::ExtendedDeterministicUniquenessValidator.install_support

ActiveRecord::Schema.define do
  create_table :users, force: :true do |t|
    t.string :thing
  end
end

class User < ActiveRecord::Base
  encrypts :thing, deterministic: true#, downcase: true
  validates :thing, uniqueness: true
end

class BugTest < Minitest::Test
  def test_association_stuff
    User.create!(thing: "my_test_string")
    model = User.create(thing: "my_test_string")
   
    puts model.errors.inspect
    assert_equal 1, model.errors.size
  end
end
```

### Expected behavior

- Single Query should be fired per encryption type for a uniqueness validation.
- Single validation error should be raised per encryption type.

### Actual behavior
**Case A - downcase: false or unset**

- Two queries will be fired when trying to validate the uniqueness of the attribute.
- Both the queries will be the same, validating encrypted strings against the database values.
- Two uniqueness validation errors will be raised.

**Case B - downcase: true**

- In this case, the number of queries will remain the same i.e. 2 but only a single validation error will be raised.
- This is because the encrypted_string is converted to downcase before comparison.

### Reason for the additional query
- This call in [extended_deterministic_uniqueness_validator](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/encryption/extended_deterministic_uniqueness_validator.rb#L20) treats the encrypted value as an original value for the attribute and converts the value as per upon the options specified (for eg: downcase)

### Possible solution

The [loop](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/encryption/extended_deterministic_uniqueness_validator.rb#L17) should be changed to only iterate over **[*encrypted_type.previous_types]**. In other words, it should exclude the current encrypted/ciphered value.

An additional query with the current encrypted/ciphered value should not be fired. I'm unable to think of any use case where this re-validation is needed. Additional queries for previous_types makes sense.

I need another set of eyes, maybe someone from the Core team or someone who has worked on the Encrypted module for confirmation. After that, I can go ahead and create a PR for the changes.

**Additional context:** https://github.com/rails/rails/issues/48248

### System configuration
**Rails version**: 7.0.4.3
**Ruby version**: 3.0.1

### PR DESCRIPTION
**Pull request by iamradioactive**
### Motivation / Background

Fixes #48445
Fixes #48248

As per the current behaviour, Two queries are being fired to validate uniqueness of an encrypted attribute even in cases when **encrypted_type.previous_types** is nil. An additional query with the current encrypted/ciphered value should not be fired. 

### Detail

This PR updates the loop in **[ExtendedDeterministicUniquenessValidator](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/encryption/extended_deterministic_uniqueness_validator.rb#L17)** to iterate over **encrypted_type.previous_types** only i.e. to exclude the current encrypted_type.

### Additional information

For Additional context Please refer to #48445

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/lib/active_record/encryption/extended_deterministic_uniqueness_validator.rb [hunk 1]
```diff
@@ -14,7 +14,7 @@ def validate_each(record, attribute, value)
           klass = record.class
           if klass.deterministic_encrypted_attributes&.include?(attribute)
             encrypted_type = klass.type_for_attribute(attribute)
-            [ encrypted_type, *encrypted_type.previous_types ].each do |type|
+            [ *encrypted_type.previous_types ].each do |type|
```
[the remark this answers is not in the document]
Comment (iamradioactive) [reply activerecord/lib/active_record/encryption/extended_deterministic_uniqueness_validator.rb]: @nvasilevski , Thanks for noticing, I missed it. I have removed the unnecessary operation!

**Review by jorgemanrubia — APPROVED** [review]
Great work tracking this one down @iamradioactive 🙏

**Comment by ghiculescu** [comment]
@iamradioactive do you mind adding `Fixes #48248` to your PR body? I confirmed this will also fix https://github.com/rails/rails/issues/48248

**Comment by nvasilevski** [comment]
> to iterate over encrypted_type.previous_types only i.e. to exclude the current encrypted_type.

Just for my own understanding, by dropping `encrypted_type` from the list, are we saying that we deliberately don't want to validate `encrypted_type` or it will still be validated because it will be included in `encrypted_type`? Thanks

The change itself looks clear!

**Review by nvasilevski — APPROVED** [review]

**Comment by iamradioactive** [comment]
> @iamradioactive do you mind adding `Fixes #48248` to your PR body? I confirmed this will also fix #48248

@ghiculescu , Seems like it's already added. Thanks @byroot

**Comment by ghiculescu** [comment]
Thanks for the fix @iamradioactive 👏

**Pull request merged**